linkerd/app/outbound/src/http/handle_proxy_error_headers.rs

```diff
@@ -0,0 +1,253 @@
+
+    #[inline]
+    fn new_service(&self, target: T) -> Self::Service {
+        let Closable(closable) = self.extract.extract_param(&target);
+        let is_meshed = matches!(
+            self.extract.extract_param(&target),
+            tls::ConditionalClientTls::Some(_)
+        );
+        let inner = self.inner.new_service(target);
+        HandleProxyErrorHeaders {
+            inner,
+            closable: closable && is_meshed,
+        }
+    }
+}
+
+impl<S, A, B> svc::Service<http::Request<A>> for HandleProxyErrorHeaders<S>
+where
+    S: svc::Service<http::Request<A>, Response = http::Response<B>>,
+    S::Response: Send,
+    S::Future: Send + 'static,
+    B: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    #[inline]
+    fn call(&mut self, req: http::Request<A>) -> Self::Future {
+        let client = req
+            .extensions()
+            .get::<ClientHandle>()
+            .cloned()
+            .expect("missing client handle");
+        let inner = self.inner.call(req);
+        ResponseFuture {
+            inner,
+            client,
+            closable: self.closable,
+        }
+    }
+}
+
+impl<B, F: TryFuture<Ok = http::Response<B>>> Future for ResponseFuture<F> {
+    type Output = Result<http::Response<B>, F::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let mut rsp = futures::ready!(this.inner.try_poll(cx))?;
+        if update_response(&mut rsp, *this.closable) {
+            // Signal that the proxy's server-side connection should be
+            // terminated. This handles the remote error as if the local proxy
+            // encountered an error.
+            debug!("Closing server connection");
+            this.client.close.close();
+        }
+        Poll::Ready(Ok(rsp))
+    }
+}
+
+fn update_response<B>(rsp: &mut http::Response<B>, closable: bool) -> bool {
+    // Clear the headers.
+    let hdr = rsp.headers_mut().remove(L5D_PROXY_CONNECTION);
+    let err = rsp.headers_mut().get(L5D_PROXY_ERROR);
+    debug!(
+        error = err
+            .and_then(|v| v.to_str().ok())
+            .map(tracing::field::display),
+        "Remote proxy error"
+    );
+
+    if !closable {
+        return false;
+    }
+
+    let is_close = hdr.map(|v| v.as_bytes() == b"close").unwrap_or(false);
+    if !is_close {
+        return false;
+    }
+
+    if rsp.version() == http::Version::HTTP_11 {
+        // If the response is HTTP/1.1, we need to send a Connection: close
+        // header to tell the application this connection is being closed.
+        rsp.headers_mut().insert(
+            http::header::CONNECTION,
+            http::HeaderValue::from_static("close"),
+        );
+    }
```

Comment:
nit, take it or leave it: can we do something like this?
```suggestion
    const CLOSE: http::HeaderValue = http::HeaderValue::from_static("close");
    
    if hdr != Some(CLOSE) {
        return false;
    }

    if rsp.version() == http::Version::HTTP_11 {
        // If the response is HTTP/1.1, we need to send a Connection: close
        // header to tell the application this connection is being closed.
        rsp.headers_mut().insert(
            http::header::CONNECTION,
            CLOSE,
        );
    }
```

Reply:
no, but we can use a static to achieve the same thing :)